### PR TITLE
count 64 bits at a time instead of 8

### PR DESCRIPTION
**Pull request by brentp**
I'm not sure if `unsafeAddr` is OK here or if there's another way. Processing 64 bits at a time makes the following code 8X faster for the first timing and 2X faster for the 2nd.

```Nim
import times


var s : set[uint16]
var t0 = cpuTime()

for i in 0'u16..uint16.high.uint16:
  s.incl(i)
  doAssert s.card == i.int + 1
echo cpuTime() - t0


s = {2'u16, 12, 32}
t0 = cpuTime()
for i in 0'u16..uint16.high.uint16:
  doAssert s.card == 3
echo cpuTime() - t0
```

### CODE REVIEW

**Comment by brentp** [comment]
the difference is 14X and 4X when using `--passC:=march=native`

File: lib/system/sets.nim [hunk 1]
```diff
@@ -30,6 +30,10 @@ proc countBits64(n: uint64): int {.compilerproc.} =
...
 
 proc cardSet(s: NimSet, len: int): int {.compilerproc, inline.} =
-  for i in 0..<len:
-    if likely(s[i] == 0): continue
+  var j = -1
+  for i in countup(0, len - 8, 8):
+    inc(result, countBits64((cast[ptr uint64](s[i].unsafeAddr))[]))
+    j = i + 7
+
+  for i in (j+1)..<len:
```
Comment (juancarlospaco) [inline lib/system/sets.nim]:
```suggestion
  for i in (j + 1) ..< len:
```

File: lib/system/sets.nim [hunk 2]
```diff
@@ -30,6 +30,10 @@ proc countBits64(n: uint64): int {.compilerproc.} =
   result = ((v * 0x0101010101010101'u64) shr 56'u64).int
 
 proc cardSet(s: NimSet, len: int): int {.compilerproc, inline.} =
-  for i in 0..<len:
-    if likely(s[i] == 0): continue
+  var j = -1
+  for i in countup(0, len - 8, 8):
+    inc(result, countBits64((cast[ptr uint64](s[i].unsafeAddr))[]))
```
Comment (krux02) [inline lib/system/sets.nim]: This is very likely to become an unaligned pointer. In your example program you are lucky and the array will be in aligned memory. You should either ensure to start at an aligned memory address here, or the type itself should change to be an array of ``uint64`` to ensure a pointer to the beginning is 8 byte aligned.

Comment (Araq) [reply lib/system/sets.nim]: On x86 unaligned are no problem though. Maybe we should make this architecture specific.

Comment (brentp) [reply lib/system/sets.nim]: so:
```Nim
proc cardSet(s: NimSet, len: int): int {.compilerproc, inline.} =
  var j = -1
  when defined(x86):
    for i in countup(0, len - 8, 8):
      inc(result, countBits64((cast[ptr uint64](s[i].unsafeAddr))[]))
      j = i + 7

  for i in (j + 1) ..< len:
    inc(result, countBits32(uint32(s[i])))
```
?

File: lib/system/sets.nim [hunk 3]
```diff
@@ -30,6 +30,11 @@ proc countBits64(n: uint64): int {.compilerproc.} =
   result = ((v * 0x0101010101010101'u64) shr 56'u64).int
 
 proc cardSet(s: NimSet, len: int): int {.compilerproc, inline.} =
-  for i in 0..<len:
-    if likely(s[i] == 0): continue
+  var j = -1
+  when defined(x86):
```
Comment (Araq) [inline lib/system/sets.nim]: ``when defined(x86) or defined(amd64)``

File: lib/system/sets.nim [hunk 4]
```diff
@@ -30,6 +30,11 @@ proc countBits64(n: uint64): int {.compilerproc.} =
   result = ((v * 0x0101010101010101'u64) shr 56'u64).int
 
 proc cardSet(s: NimSet, len: int): int {.compilerproc, inline.} =
-  for i in 0..<len:
-    if likely(s[i] == 0): continue
+  var j = -1
+  when defined(x86) or defined(amd64):
+    for i in countup(0, len - 8, 8):
```
Comment (Araq) [inline lib/system/sets.nim]: Convoluted. Use a while loop so that you can use the index after the loop.

**Pull request merged**